From a07d22ce4456aa4e5a249097fefe11665638899a Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hauke.mehrtens@intel.com>
Date: Tue, 30 Jun 2020 16:59:21 +0200
Subject: [PATCH] PONRTSYS-7716: Allow unlinking of inactive queues

If we create a queue with dp_node_alloc() and link it successfully with
dp_node_link_add() the unlink with dp_node_unlink() will fail in case
no scheduler  was added in between.

This fixes the problem by allowing to unlink queues which are not
active.

This patch was provided by Guohua
---
 drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c | 17 ++++++++++-------
 1 file changed, 10 insertions(+), 7 deletions(-)

diff --git a/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c b/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c
index 568010a0bb96..683c0a8bb21b 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_ppv4_api.c
@@ -2738,10 +2738,10 @@ int dp_node_unlink_31(struct dp_node_link *info, int flag)
 		node_id = qos_queue_stat->node_id;
 		/* Need to check ACTIVE Flag */
 		if (!(qos_queue_stat->flag & PP_NODE_ACTIVE)) {
-			pr_err("Wrong Queue[%d] Stat(%d):Expect ACTIVE\n",
-			       info->node_id.q_id,
-			       qos_queue_stat->flag);
-			return DP_FAILURE;
+			DP_DEBUG(DP_DBG_FLAG_QOS,
+			       "Queue[%d]:No need unlink since not active\n",
+			       info->node_id.q_id);
+			return DP_SUCCESS;
 		}
 		cqm_qid2ep_map_set(info->node_id.q_id, priv->ppv4_drop_p);
 		DP_DEBUG(DP_DBG_FLAG_QOS, "%s qid:%d, dq_port:%d\n",
@@ -2750,13 +2750,16 @@ int dp_node_unlink_31(struct dp_node_link *info, int flag)
 		queue_flush_31(info->inst, node_id, 0);
 	} else if (info->node_type == DP_NODE_SCH) {
 		if (!(qos_sch_stat->c_flag & PP_NODE_ACTIVE)) {
-			pr_err("Wrong Sched FLAG Expect ACTIVE\n");
-			return DP_FAILURE;
+			DP_DEBUG(DP_DBG_FLAG_QOS,
+				 "Sched[%d]: No need unlink since not active\n",
+				 info->node_id.sch_id);
+			return DP_SUCCESS;
 		}
 
 		if (dp_qos_sched_get_queues(priv->qdev, info->node_id.sch_id,
 					    queue_buf, queue_size, &queue_num))
-			return DP_FAILURE;
+			return DP_SUCCESS;
+
 		for (i = 0; i < queue_num; i++) {
 			if (!(priv->qos_queue_stat[queue_buf[i]].flag &
 			      PP_NODE_ACTIVE))
